code/scripts/espurna_utils/bsec.py

```diff
@@ -0,0 +1,107 @@
+import os
+
+def bsec_inject_patcher(env):
+    platform = env.PioPlatform()
+    framework_dir = platform.get_package_dir("framework-arduinoespressif8266")
+    bsec_dir = os.path.join(
+        "${PROJECT_LIBDEPS_DIR}", "${PIOENV}", "BSEC Software Library", "src", "esp8266"
+    )
+
+    env.Append(LIBPATH=["${PROJECT_LIBDEPS_DIR}/${PIOENV}/BSEC Software Library/src/esp8266"])
+    env.Append(LIBS=["algobsec"])
+
+    patchflag_path = os.path.join(env.subst(bsec_dir), ".patching-done")
+
+    if os.path.isfile(patchflag_path):
+      print("Archive libalgosec.a has already been patched")
+      return
+
+    toolchain_ar = os.path.join(
+        platform.get_package_dir("toolchain-xtensa"), "bin", "xtensa-lx106-elf-ar"
+    )
+
+    # Backup libalgobsec.a into libalgobsec.a.orig.
+    backup_action = env.VerboseAction(
+        " ".join(
+            [
+               '/bin/mv',
+                os.path.join(bsec_dir.replace(' ', '\\ '), "libalgobsec.a"),
+                os.path.join(bsec_dir.replace(' ', '\\ '), "libalgobsec.a.orig")
+            ]
+        ),
+        "Backing up BSEC libalgosec.a to libalgosec.a.orig",
+    )
+
+    # Extract libalgobsec.a into *.o files.
+    extract_action = env.VerboseAction(
+        " ".join(
+            [
+               toolchain_ar,
+                "-xv",
+                os.path.join(bsec_dir.replace(' ', '\\ '), "libalgobsec.a.orig")
+            ]
+        ),
+        "Extracting BSEC libalgobsec.a.orig",
+    )
+
+    # Archive *.c.o files into libalgobsec.a.
+    archive_action = env.VerboseAction(
+        " ".join(
+            [
+               toolchain_ar,
+                "-crv",
+                os.path.join(bsec_dir.replace(' ', '\\ '), "libalgobsec.a"),
+                "*.c.o"
+            ]
+        ),
+        "Archiving BSEC libalgosec.a",
+    )
+
+    # Create patch flag file.
+    touch_action = env.VerboseAction(
+        " ".join(
+            [
+               "/usr/bin/touch",
+                patchflag_path.replace(' ', '\\ '),
+            ]
+        ),
+        "Touching patch flag",
+    )
+
+    # Remove renamed files.
+    cleanup_action = env.VerboseAction(
+        " ".join(
+            [
+               "rm",
+                "*.c.o",
+            ]
+        ),
+        "Cleaning up",
+    )
+
+    env.Execute(backup_action)
+    env.Execute(extract_action)
+
+    rename_files()
```

Comment:
@mcspr can this be converted into a `VerboseAction` as well? I've had issues running more complex shell commands.